burstphoto/align.metal

```diff
@@ -209,6 +485,84 @@ kernel void compute_tile_alignments(texture3d<float, access::read> tile_diff [[t
+                                     uint2 gid [[thread_position_in_grid]]) {
+    
+    // load args
+    int texture_width = ref_texture.get_width();
+    int texture_height = ref_texture.get_height();
+    int tile_half_size = tile_size / 2;
+    
+    // initialize some variables
+    int3 comp_tile_x, comp_tile_y;
+    float ref_val;
+    
+    // compute tile position if previous alignment were 0
+    int x0 = int(floor( tile_half_size + float(gid.x)/float(n_tiles_x-1) * (texture_width  - tile_size - 1) ));
+    int y0 = int(floor( tile_half_size + float(gid.y)/float(n_tiles_y-1) * (texture_height - tile_size - 1) ));
+
+    // calculate shifts of gid index for 3 candidate alignments to evaluate
+    int3 const x_shift = int3(0, ((gid.x%2 == 0) ? -1 : 1), 0);
+    int3 const y_shift = int3(0, 0, ((gid.y%2 == 0) ? -1 : 1));
+    
+    int3 const x = clamp(int3(gid.x+x_shift), 0, n_tiles_x-1);
+    int3 const y = clamp(int3(gid.y+y_shift), 0, n_tiles_y-1);
+    
+    // factor in previous alignment for 3 candidates
+    int4 prev_align0 = prev_alignment.read(uint2(x[0], y[0]));
+    int4 prev_align1 = prev_alignment.read(uint2(x[1], y[1]));
+    int4 prev_align2 = prev_alignment.read(uint2(x[2], y[2]));
+    
+    int3 dx0 = downscale_factor * int3(prev_align0.x, prev_align1.x, prev_align2.x);
+    int3 dy0 = downscale_factor * int3(prev_align0.y, prev_align1.y, prev_align2.y);
+     
+    // compute tile differences for 3 candidates
+    float diff[3] = {0.0f, 0.0f, 0.0f};
+    
+    for (int dx1 = -tile_half_size; dx1 < tile_half_size; dx1++) {
+        for (int dy1 = -tile_half_size; dy1 < tile_half_size; dy1++) {
+            
+            // compute the indices of the pixels to compare
+            int const ref_tile_x = x0 + dx1;
+            int const ref_tile_y = y0 + dy1;
+            ref_val = ref_texture.read(uint2(ref_tile_x, ref_tile_y)).r;
+             
+            comp_tile_x = ref_tile_x + dx0;
+            comp_tile_y = ref_tile_y + dy0;
+            
+            // check 3 candidates
+            for (int i = 0; i < 3; i++) {
+            
+                if ((comp_tile_x[i] < 0) || (comp_tile_y[i] < 0) || (comp_tile_x[i] >= texture_width) || (comp_tile_y[i] >= texture_height)) {
+                    diff[i] += abs(ref_val + 2*UINT16_MAX_VAL);
+                } else {
+                    diff[i] += abs(ref_val - comp_texture.read(uint2(comp_tile_x[i], comp_tile_y[i])).r);
+                }
+            }
+        }
+    }
+    
+    // store corrected (best) alignment
+    if(diff[0] < diff[1] & diff[0] < diff[2]) {
+        prev_alignment_corrected.write(prev_align0, gid);
+        
+    } else if(diff[1] < diff[2]) {
+        prev_alignment_corrected.write(prev_align1, gid);
+        
+    } else {
+        prev_alignment_corrected.write(prev_align2, gid);
+    }
+}
+
+
 kernel void warp_texture(texture2d<float, access::read> in_texture [[texture(0)]],
```

Comment:
Sams as for the others. Does get_width and get_height have to be called inside this function? To the (limited) extent that I understand how metal works, I am under the impression that this will get called an incredible amount of time in parallel. And I don't know if this call gets compiled out of those parallel calls.

Reply:
I think Metal is extremely efficient in extracting this information from the texture. Still it can be checked in a future version if performance can be increased.

Reply:
This suggestion does not show any improvement of performance.